samples/extensions/khr/externalmemory/main.cpp

```diff
@@ -0,0 +1,635 @@
+        queue_create_info.sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
+        queue_create_info.queueFamilyIndex = 0;
+        queue_create_info.queueCount = 1;
+        queue_create_info.pQueuePriorities = &default_queue_priority;
+
+        VkDeviceCreateInfo device_create_info{};
+        device_create_info.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
+        device_create_info.queueCreateInfoCount = 1;
+        device_create_info.pQueueCreateInfos = &queue_create_info;
+        device_create_info.enabledExtensionCount =
+            static_cast<uint32_t>(required_device_extensions.size());
+        device_create_info.ppEnabledExtensionNames =
+            required_device_extensions.data();
+
+        VkDevice vk_device;
+        VK_CHECK(vkCreateDevice(vk_physical_device, &device_create_info,
+                                nullptr, &vk_device));
+
+        if (!diag_opts.quiet)
+        {
+            std::cout << "Selected platform: "
+                      << cl_platform.getInfo<CL_PLATFORM_VENDOR>() << "\n"
+                      << "Selected device: "
+                      << cl_device.getInfo<CL_DEVICE_NAME>() << "\n"
+                      << std::endl;
+        }
+
+        // Create OpenCL runtime objects.
+        cl::Context cl_context{ cl_device };
+
+        // Check if the device supports the Khronos extensions needed before
+        // attempting to compile the kernel.
+        if (diag_opts.verbose)
+        {
+            std::cout << "Checking Khronos extensions support... ";
+            std::cout.flush();
+        }
+
+        for (const auto& extension : required_khronos_extensions)
+        {
+            if (!cl::util::supports_extension(cl_device, extension))
+            {
+                std::cout << "OpenCL device does not support the required "
+                             "Khronos extension "
+                          << extension << std::endl;
+                exit(EXIT_SUCCESS);
+            }
+        }
+
+        // Compile kernel.
+        if (diag_opts.verbose)
+        {
+            std::cout << " done.\nCompiling OpenCL kernel... ";
+            std::cout.flush();
+        }
+        const char* kernel_location = "./external_saxpy.cl";
+        std::ifstream kernel_stream{ kernel_location };
+        if (!kernel_stream.is_open())
+            throw std::runtime_error{
+                std::string{ "Cannot open kernel source: " } + kernel_location
+            };
+        cl::Program cl_program{
+            cl_context,
+            std::string{ std::istreambuf_iterator<char>{ kernel_stream },
+                         std::istreambuf_iterator<char>{} }
+        };
+
+        // The Khronos extension showcased requires OpenCL 3.0 version.
+        cl::string compiler_options = "";
+        const std::string dev_version =
+            cl_device.getInfo<CL_DEVICE_OPENCL_C_VERSION>();
+        if (opencl_version_contains(dev_version, "3."))
```

Comment:
This isn't the right way to check for this.  [CL_DEVICE_OPENCL_C_VERSION](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_OPENCL_C_VERSION) is the older query that returns the "highest fully backwards compatible OpenCL C version supported by the compiler for the device".  This means that many OpenCL 3.0 devices will only return "OpenCL 1.2" for this query.

The right way to query for OpenCL C 3.0 support is via [CL_DEVICE_OPENCL_C_ALL_VERSIONS](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_OPENCL_C_ALL_VERSIONS).

This really ought to be wrapped in an SDK utility function.

Reply:
Arg yes you're right, fixed this now.

Btw, I checked whether this was used in other examples and found out it was also used in the [reduce](https://github.com/KhronosGroup/OpenCL-SDK/blob/main/samples/core/reduce/main.c#L148) core sample. I think it should be changed there as well.